### PR TITLE
Tell consumers which props were modified

### PR DESCRIPTION
**Pull request by trusktr**
Builds off of PR https://github.com/skatejs/skatejs/pull/1508. Related issue: https://github.com/skatejs/skatejs/issues/1489

* [ ] Bug
* [x] Feature

## Requirements

* [x] Read the [contribution guidelines](https://github.com/skatejs/skatejs/blob/5.x/CONTRIBUTING.md).
* [ ] Wrote tests.
* [ ] Updated docs and upgrade instructions, if necessary.

## Rationale

SkateJS already knows which props have changed, so we can drastically clean up end user code by simply exposing this information.

This drastically cleaned up some of my code. For example, I had code that I changed from

```js
    constructor(element) {
        Super(this).constructor(element)

        Private(this).lastSize = [null, null, null]
        Private(this).lastSizeMode = [null, null, null]
    },

    updated( oldProps, oldState ) {
        if (Private(this).sizeChanged(this.size, this.sizeMode)) {
            this._updateGeometryOnSizeChange(this.size)
        }
    },

    private: {
        lastSize: [null, null, null],
        lastSizeMode: [null, null, null],

        sizeChanged(newSize, newSizeMode) {
            let result = false

            if (
                this.lastSize[0] !== newSize.x ||
                this.lastSize[1] !== newSize.y ||
                this.lastSize[2] !== newSize.z
            ) {
                result = true
                this.lastSize[0] = newSize.x
                this.lastSize[1] = newSize.y
                this.lastSize[2] = newSize.z
            }

            if (
                this.lastSizeMode[0] !== newSizeMode.x ||
                this.lastSizeMode[1] !== newSizeMode.y ||
                this.lastSizeMode[2] !== newSizeMode.z
            ) {
                result = true
                this.lastSizeMode[0] = newSizeMode.x
                this.lastSizeMode[1] = newSizeMode.y
                this.lastSizeMode[2] = newSizeMode.z
            }

            return result
        },
    },
```

to

```js
    updated( oldProps, oldState, modifiedProps ) {
        if ( modifiedProps.has('size') || modifiedProps.has('sizeMode') ) {
            this._updateGeometryOnSizeChange(this.size)
        }
    },
```

This is highly valuable!

## Implementation

Implemented this way to be backwards compatible. Existing code continues to work as is.

## Tasks

_List any tasks you need to do here, if any. Delete this section if you don't need it._

* [ ] test
* [ ] docs

### CODE REVIEW

**Comment by matthewp** [comment]
Yeah, I agree with @blikblum here. Is this not possible to build as a separate mixin?

File: packages/skatejs/src/with-update.js [hunk 1]
```diff
@@ -138,6 +139,7 @@ export const withUpdate = (Base: Class<any> = HTMLElement): Class<any> =>
     _prevState: Object;
     _props: Object;
     _state: Object;
+    _modifiedProps: Set;
```
Comment (treshugart) [inline packages/skatejs/src/with-update.js]: Why not just use an object?

File: packages/skatejs/src/with-update.js [hunk 2]
```diff
@@ -241,6 +245,9 @@ export const withUpdate = (Base: Class<any> = HTMLElement): Class<any> =>
       if (super.connectedCallback) {
         super.connectedCallback();
       }
+      keys(this.constructor.props).forEach(prop => {
```
Comment (treshugart) [inline packages/skatejs/src/with-update.js]: Why do all props need to be signalled as changing when connected?

Comment (trusktr) [reply packages/skatejs/src/with-update.js]: This would be effectively the same as firing a default `attributeChangedCallback` on connecting, so my thought here was that we want to react to all the values. Usually we need the values for something, so we need to use them all initially. In my cases for example, I need to set WebGL rendering properties/features based on the initial values.

There isn't a way to know if `updated` is being called for the firs time, and the `updated` method will generally always contain conditional checks to determine which props changed, so in order to fire logic for all props, this is why I thought to initially set all props modified.

If we don't set them all as "modified" initially, then the only option we have is to avoid the use of `modifiedProps` in the `updated` method and instead diff `oldProps` vs `this.props`, which defeats the purpose of having `modifiedProps`.

Is there maybe a better way to run initial logic for all props?

**Comment by trusktr** [comment]
> Is this not possible to build as a separate mixin?

@matthewp  It doesn't seem easy to do this without duplicating much of the efforts already put into `withUpdate`.

> This will incurs in performance penalty (plus increased code size and memory consumption) for majority of cases / users that don't need this feature.

@blikblum Any thoughts on how we might improve it, maybe a way for classes to opt-in?

For big applications with lots of data/state, I'm fairly certain the cost of maintaining a `Set<string>` will be more efficient than repeatedly creating/destroying/diffing objects.

> Also breaks consistency between state/props (i would expect state have the same feature)

State's already completely different: it's basically a single prop whose value is always an object. Personally I never use it, props have all the reactivity I need. This is similar to Vue components (they don't have a `state` object, just props).

We could remove state, and effectively have the same thing by defining it with props:

```js
static props = {
  state: props.object,
}
```

@treshugart 

> Why not just use an object?

Hmmmm... do you mean, like have the keys on an object so they are constantly there, as booleans? Then

```js
updated(oldProps, oldState, changedProps) {
  if (changedProps.someProp) console.log('someProp changed!')
}
```

That's a good idea! Then compared to adding/removing string from a set the memory use doesn't fluctuate, and lookup is faster!

I'll update...

**Comment by trusktr** [comment]
Alright, I've updated the PR. It seems like this would be cleaner.

The `connectedCallback` initially adds all the props to the _modifiedProps object (therefore updated logic can react to initial prop values).

Checking boolean values is about 5.5 times faster than checking for values in a Set according to this simple test in my Chrome:

```js
let o = {n:1}
let s = new Set([1,2,3,4])

var start = performance.now()
for (let i=0; i<5000000; i++) o.n
var end = performance.now()
console.log(end-start)

var start = performance.now()
for (let i=0; i<5000000; i++) s.has(4)
var end = performance.now()
console.log(end-start)
```

However, the `for..in` loop to clear reset the `_modifiedProps` object seems like it may be slower than clearing a Set.

**Comment by trusktr** [comment]
I just did some perf tests.

- `for..in` takes 1/2 the time of `keys(o).forEach`.
  - iterating over 15 props with `for..in` 5,000,000 times took ~2300ms.
- `Object.keys(o).forEach` takes 3/5ths the time of `for..in`
  - iterating over 15 props with `Object.keys(o).forEach` 5,000,000 times took ~1550ms.
- `Object.keys(o)` paired with a plain `for (;;)` takes 1/31th the time of `Object.keys(o).forEach`
  - iterating over 15 props with the plain loop 5,000,000 times took ~50ms.

Updated the PR to use plain loops and to only ever call `Object.keys` one time.

File: packages/skatejs/src/with-update.js [hunk 3]
```diff
@@ -133,11 +135,13 @@ export const withUpdate = (Base: Class<any> = HTMLElement): Class<any> =>
     static _attributeToPropertyMap: Object;
     static _observedAttributes: Array<string>;
     static _props: Object;
+    static _propsList: Array<string>;
```
Comment (trusktr) [inline packages/skatejs/src/with-update.js]: I made `_propsList` `static` instead of an instance variable so there's only one instance.

**Comment by trusktr** [comment]
This implementation is fast now, and lightweight compared to diffing possibly much data across possibly many component instances.

It is currently branched from a branch of an unrelated feature, let me rebase it on master because the feature is standalone from my other pull requests.

**Comment by trusktr** [comment]
Alright, I updated this PR to have `master` as base.

**Comment by treshugart** [comment]
`@skatejs/element` already does this. As discussed this does it by using an object of old values. We've identified that this could, in theory, hinder 60fps performance but don't seem to have proven it yet. Since the discussion here is very long and hard to parse, we haven't proven a performance problem and it's already in a release (and it'd be a breaking change to change it to this), I'd prefer we:

1. Prove it's a problem.
2. Flesh it out in a separate issue keeping the discussion focused and terse.
